### PR TITLE
Hardware keys support in Connect POC

### PR DESCRIPTION
**Pull request by gzdunek**
This PR is a POC of the hardware key support in Connect. The code is not very polished, I wanted to discuss my concerns before going too far. After the final solution comes up, I will close this PR and open smaller ones.
The first two commits are from Brian, probably you can skip them for now (that's the part I trust the most :) ).

### Overview
To make hardware keys work in Connect we need two things:
- Allowing concurrent use of the hardware key. By design, the key can only be opened by a single handler at a time, which caused different parts of tshd to compete for access. My first idea was to add a global lock on the key, but this resulted in quite bad performance (every opening of the physical key takes 10-20 ms). Fortunately, Brian already had a better fix for the problem (caching a connection to the key for a few seconds), so I just rebased onto his branch. One thing that I will have to add it is handling the key disconnection while the connection is cached (now it causes panics).
- Providing a prompt interface to `yubikey.go` that would allow tshd pass a custom implementation instead of the terminal `prompt` methods. I came up with an interface with four methods: `AskPIN`, `Touch`, `ChangePIN` and `ConformSlotOverwrite`. 
These prompts are displayed as important dialogs on the Electron side. The UX of this is not perfect, since if you have to unlock the hardware key during log-in, you will see a dialog on top of other dialog. Earlier I considered integrating these prompts with a login dialog, but that would require some significant changes on the frontend side (if there’s a login in progress the prompt should be a part of the login dialog, otherwise it should be a standalone dialog). Because of the additional complexity, I will keep this out of scope. 

### Problems
My main issue is with prompting for PIN/touch while establishing the gRPC connection (I think Nic saw it coming :D):
- There’s a 30 seconds dial-in timeout. The interaction between the code that performs the TLS handshake and the hardware key happens through a `singer.Sign` interface. Unfortunately, this interface doesn't provide `Context` (I think it wasn't designed to be async), so I have no information if a context of the TLS handshake was cancelled. Because of that, in the app you can see that some calls failed in the background (because of a timeout, for example), but the prompt modal is still visible. This is not ideal, but I think we can live with it.
- A bigger issue arises when the user cancels the prompt or the timeout occurs. In such cases, the TLS handshake fails, causing the gRPC connection to enter a [`TRANSIENT_FAILURE`](https://grpc.github.io/grpc/core/md_doc_connectivity-semantics-and-api.html) state where it tries to reconnect with an exponential backoff. This creates a poor user experience, as each time the user clicks "Cancel," a new prompt modal keeps appearing  :/ In theory, I could change the backoff values to basically disable it, but that's probably a useful mechanism for bad network conditions.
- Another thing is that even with the backoff disabled, the connection does not behave as I would like it to. If I send a new RPC, it immediately fails, even without trying to reconnect. I know there is an option `grpc.waitForReady`, but it makes the connection wait indefinitely, without reconnecting.

What I would like to achieve is to make the connection go to an IDLE state when the handshake fails and then reconnect only when a new RPC arrives. This would result in an optimal UX, when the user is prompted for pin/touch only if they performed an action.
- A thing that I considered, was closing the entire cluster client in some gRPC middleware when I receive an error saying that the user closed the hardware key modal. This is a quite aggressive approach 😓, but I don’t have any better idea at the moment.

An alternative solution to all these problems would be to make a random one-off signature before creating the gRPC connection (this would unlock the key for a few seconds allowing the gRPC connection to be created without a prompt).
Let me quote Nic:
>[…] I could see us fighting with gRPC trying to make a bunch of TLS dials that time out before the user can input the pin. Feels like something where eventually we might have to proactively make sure the yubikey is ready before handing off to gRPC and waiting for a pin in the middle of a TLS handshake, but that seems really tricky

But when to do it? What if the gRPC connection goes down while the app is running and needs to be re-created?
Brian also pointed out that this would ruin the elegance of treating YubiKey as any other private key signer :(

I’d appreciate any advice on how to proceed. I’d like to keep the current approach of requesting a PIN/touch during the gRPC connection setup. However, due to my limited understanding of gRPC, TLS, and our dialers, I’m unsure about what’s doable and what isn’t.

### How to test it
1. Build tsh with PIV: `PIV=dynamic make build/tsh`
2. Require hardware key in your role https://goteleport.com/docs/admin-guides/access-controls/guides/hardware-key-support/#step-12-enforce-hardware-key-support (I set it to `hardware_key_touch_and_pin`).
3. If you haven't created PIV PIN before, please do it with `tsh login`, I haven't implemented it yet in the app.

### Quick demo
A happy path:

https://github.com/user-attachments/assets/383da9c6-94a4-44df-92b6-593560ca7c41

And a case where the prompt opens over and over again:

https://github.com/user-attachments/assets/3c793db2-6632-449f-a5cb-1a75f4198fe2

Uh, that's the longest PR description I've ever written!

### CODE REVIEW

File: api/utils/keys/privatekey.go [hunk 1]
```diff
@@ -214,6 +214,12 @@ func LoadPrivateKey(keyFile string) (*PrivateKey, error) {
 
 // ParsePrivateKey returns the PrivateKey for the given key PEM block.
 func ParsePrivateKey(keyPEM []byte) (*PrivateKey, error) {
+	//TODO(gzdunek): ParsePrivateKey has 60 usages, so I created ParsePrivateKeyWithPrompt
+	return ParsePrivateKeyWithPrompt(keyPEM, nil)
```
Comment (gzdunek) [inline api/utils/keys/privatekey.go:218]: Passing the `HardwareKeyPrompt` through so many layers is painful, especially in cases like this one, where `ParsePrivateKey` has so many usages. To make my life easier I created a new version of this function, to allow it specify a custom prompt. 
If the prompt is `nil`, I use the default `CLIPrompt`. I know it's a bit defensive technique, but idk if adjusting all the callsites would be better.

**Comment by nklaassen** [comment]
Hmm... the dialer is probably the wrong place actually. But maybe where we fetch the creds, or if this isn't right we could implement our own grpc/credentials.TransportCredentials
```diff
diff --git a/api/client/proxy/client.go b/api/client/proxy/client.go
index 817da42892..a66c1a2989 100644
--- a/api/client/proxy/client.go
+++ b/api/client/proxy/client.go
@@ -115,6 +119,15 @@ func (c *ClientConfig) CheckAndSetDefaults() error {
                        // not to send the client certificate by looking at certificate request.
                        if len(tlsCfg.Certificates) > 0 {
                                cert := tlsCfg.Certificates[0]
+
+                               if priv, ok := cert.PrivateKey.(*keys.PrivateKey); ok && priv.IsHardware() {
+                                       b := make([]byte, 256)
+                                       _, err := priv.Sign(rand.Reader, b, crypto.SHA256)
+                                       if err != nil {
+                                               slog.Warn("failed to warm up hardware key for TLS dial")
+                                       }
+                               }
+
                                tlsCfg.Certificates = nil
                                tlsCfg.GetClientCertificate = func(_ *tls.CertificateRequestInfo) (*tls.Certificate, error) {
                                        return &cert, nil
```
again, just an idea

**Comment by Joerger** [comment]
This is a tough problem, the direct solution would require us to trace every TLS handshake path and find a way to add context, so we can pause, wait, cancel, etc... This sounds like a huge load of work if it's even possible. Maybe there is a clever solution, like using a context dialer, but it's unlikely that one clever solution alone will be enough.

Another much simpler solution would be to cache the Yubikey Connection for much longer, and handle the PIN prompt before we'd ever be in a TLS handshake. Simplest solution that comes to mind:

1. When Teleport Connect is opened or the user logs in, warm up the YubiKey connection, prompt for PIN preemptively with existing `CheckSetPin` method. Keep the connection open for some time (15 minutes?), with a handler to catch when the connection is closed.
2. When the YubiKey connection is closed, re-open it and prompt for PIN.

In both the above cases, ideally we would prompt on some user action instead. Otherwise we'll just prompt every 15 minutes. Maybe this just needs to be done whenever making a `tshd` request? Some kind of retry-with-pin-prompt mechanism?

This is probably going to be much better UX than the "clever" fix anyways. Holding the connection open for some time would also be useful for some `tsh proxy` requests that have repeated handshakes.

**Comment by nklaassen** [comment]
You make good points Brian, if we can get away with holding the yubikey connection open much longer and prompting after a user interaction, that would be great.

But in case that's not an option, I think some clever tricks could help. It wouldn't have to be exhaustive, if we ever get to a Sign and the connection isn't already open, worst case we're back to the situation we have now where we have to do our best to open it. If it's specifically the gRPC dialing that causes most problems, I think we could improve that by opening the connection right before the dial

**Comment by gzdunek** [comment]
I just tested the Nic's clever fix and it looks like it solves all my problems! I can wait now for the pin/touch as long as I want before the gRPC dialing, also I don't get these unwanted retries when the prompt is canceled (I only need to return early instead of printing a warning).
As for where to do the warm-up, it seems that `client/proxy/client.go` alone is enough? At least, from my testing looks like this the code that is called when Connect creates a gRPC connection (and as Nic said, we can treat this mechanism as a "best effort", any other place that doesn't use it will just ask for pin/touch later).

-------------
>This is probably going to be much better UX than the "clever" fix anyways. Holding the connection open for some time would also be useful for some tsh proxy requests that have repeated handshakes.

> Edit: my main concern with this solution is that holding the YubiKey PIV connection open will lock out other PIV clients, including tsh and tctl. Holding the connection open for just 1 minute and prompting only on user action would go a long way to avoiding this issue, or at least updating the error message to indicate that Teleport Connect holds the exclusive connection.

Holding the connection for the entire app lifetime would provide the best user experience. However, keeping the connection locked for more than a few seconds isn't feasible at the moment. Doing so would block not only `tsh` and `tctl` from outside Teleport Connect but also block SSHing into a server in the app (since we spawn `tsh ssh` under the hood). Even holding the connection for just a minute would be annoying—I'm trying to imagine logging in and then having to wait a full minute before I can SSH to the server 💀

>Another clever solution would be to use tshd in tsh and tctl so that the same PIV connection can be re-used, but this seems like more work than it's worth.

Yeah, that would solve the problem above, but I agree that's too much work :(

**Pull request closed without merging**